src/chkstat.cpp

```diff
@@ -21,1365 +21,1246 @@
-    }
-    else if (access(central_etc_perms.c_str(), R_OK) == 0)
-    {
-        ensure_array((void**)&permfiles, &npermfiles);
-        permfiles[npermfiles++] = strdup(central_etc_perms.c_str());
+        // TODO: anything interesting here? probably ENOENT.
+        return;
     }
 
-    // 2. central easy, secure paranoid as those are defined by SUSE
-    for (i = 0; i < nlevel; ++i)
+    struct dirent* entry = nullptr;
+    // First collect a sorted set of base files, skipping unwanted files like
+    // backup files or specific profile files. Therefore filter out duplicates
+    // using the sorted set.
+    std::set<std::string> files;
+
+    // TODO: error situations in readdir() are currently not recognized.
+    // Consequences?
+    while ((entry = readdir(dir_handle)))
     {
-        if (!strcmp(level[i], "easy")
-                || !strcmp(level[i], "secure")
-                || !strcmp(level[i], "paranoid"))
+        std::string_view name(entry->d_name);
+
+        if (name == "." || name == "..")
+            continue;
+
+        bool found_suffix = false;
+
+        /* filter out backup files */
+        for (const auto &suffix: {"~", ".rpmnew", ".rpmsave"})
         {
-            auto base = std::string("/permissions.") + level[i];
-            for( const auto &dir: { usr_root, etc_root } )
+            if (hasSuffix(name, suffix))
             {
-                std::string profile = dir + base;
-
-                if (access(profile.c_str(), R_OK) == 0)
-                {
-                    ensure_array((void**)&permfiles, &npermfiles);
-                    permfiles[npermfiles++] = strdup(profile.c_str());
-                    break;
-                }
+                found_suffix = true;
+                break;
             }
         }
+
+        if (found_suffix)
+            continue;
+
+        files.insert(std::string(name));
     }
 
-    // 3. package specific permissions
-    read_permissions_d((usr_root + "/permissions.d").c_str());
-    read_permissions_d((etc_root + "/permissions.d").c_str());
+    (void)closedir(dir_handle);
 
-    // 4. central permissions files with user defined level incl 'local'
-    for (i = 0; i < nlevel; ++i)
+    // now add the sorted set of files to the profile paths to process later
+    // on
+    for (const auto &file: files)
     {
-        if (!strcmp(level[i], "easy") || !strcmp(level[i], "secure") || !strcmp(level[i], "paranoid"))
+        if (file.find_first_of('.') != file.npos)
+            // we're only interested in base profiles
             continue;
 
-        std::string profile = etc_root + "/permissions." + level[i];
+        const auto path = dir + "/" + file;
 
-        if (access(profile.c_str(), R_OK) == 0)
+        // NOTE: we already know that the path exist(ed), because it was
+        // returned from the readdir() above. The existence check is racy
+        // anyways so no need to recheck.
```

Comment:
Thinking twice about this I think it is actually a tad inconsequential, because the original check also tested for readability while the current code implies this. Maybe we should still perform an `existsFile()` check? It would probably only matter when running `chkstat` as an unprivileged user or if file permissions are messed up.

Reply:
Indeed, usability slightly decreases in that case (as unlikely as it is). I don't think we should fret too much about the unprivileged use-case, so I leave it up to you to either add back this check or remove the comment that refers to a non-existent existence check.

Reply:
Okay I'll remove the comment and accept the slight change in behaviour here.